File: LeetCode/1195-fizz-buzz-multithreaded.java
import java.util.concurrent.Semaphore;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class FizzBuzz {
  private int n;
  private final Semaphore fSem = new Semaphore(0);
  private final Semaphore bSem = new Semaphore(0);
  private final Semaphore fbSem = new Semaphore(0);
  private final Semaphore nSem = new Semaphore(1);
  private boolean done = false;

  public FizzBuzz(int n) {
    this.n = n;
  }

  // printFizz.run() outputs "fizz".
  public void fizz(Runnable printFizz) throws InterruptedException {
    while (!done) {
      fSem.acquire();
      if (!done)
        printFizz.run();
      nSem.release();
    }
  }

  // printBuzz.run() outputs "buzz".
  public void buzz(Runnable printBuzz) throws InterruptedException {
    while (!done) {
      bSem.acquire();
      if (!done)
        printBuzz.run();
      nSem.release();
    }
  }

  // printFizzBuzz.run() outputs "fizzbuzz".
  public void fizzbuzz(Runnable printFizzBuzz) throws InterruptedException {
    while (!done) {
      fbSem.acquire();
      if (!done)
        printFizzBuzz.run();
      nSem.release();
    }
  }

  // printNumber.accept(x) outputs "x", where x is an integer.
  public void number(IntConsumer printNumber) throws InterruptedException {
    for (int i = 1; i <= this.n; i++) {
      nSem.acquire();
      if (i % 15 == 0) {
        fbSem.release();
      } else if (i % 3 == 0) {
        fSem.release();
      } else if (i % 5 == 0) {
        bSem.release();
      } else {
        if (!done)
          printNumber.accept(i);
        nSem.release();
      }
    }
    nSem.acquire();
    done = true;
    fSem.release();
    bSem.release();
    fbSem.release();
    nSem.release();
  }
}

File: LeetCode/1195-fizz-buzz-multithreaded.py
class FizzBuzz:
    def __init__(self, n: int):
        self.n = n
        self.fSem = threading.Semaphore(0)
        self.bSem = threading.Semaphore(0)
        self.fbSem = threading.Semaphore(0)
        self.nSem = threading.Semaphore(1)
        self.done = False        

    # printFizz() outputs "fizz"
    def fizz(self, printFizz: 'Callable[[], None]') -> None:
        while True:
            self.fSem.acquire()
            if self.done: break
            printFizz()
            self.nSem.release()

    # printBuzz() outputs "buzz"
    def buzz(self, printBuzz: 'Callable[[], None]') -> None:
        while True:
            self.bSem.acquire()
            if self.done: break
            printBuzz()
            self.nSem.release()

    # printFizzBuzz() outputs "fizzbuzz"
    def fizzbuzz(self, printFizzBuzz: 'Callable[[], None]') -> None:
        while True:
            self.fbSem.acquire()
            if self.done: break
            printFizzBuzz()
            self.nSem.release()

    # printNumber(x) outputs "x", where x is an integer.
    def number(self, printNumber: 'Callable[[int], None]') -> None:
        for i in range(1, self.n+1):
            self.nSem.acquire()
            if i % 15 == 0:
                self.fbSem.release()
            elif i % 3 == 0:
                self.fSem.release()
            elif i % 5 == 0:
                self.bSem.release()
            else:
                printNumber(i)
                self.nSem.release()
        self.nSem.acquire() 
        self.done = True
        self.fSem.release()
        self.bSem.release()
        self.fbSem.release()
        self.nSem.release()
